--- a/kernel/fork.c	2012-02-23 01:43:29.000000000 -0700
+++ b/kernel/fork.c	2012-02-23 08:46:50.000000000 -0700
@@ -36,7 +36,9 @@
 #include <linux/syscalls.h>
 #include <linux/jiffies.h>
 #include <linux/futex.h>
+#ifndef __GENKSYMS__
 #include <linux/compat.h>
+#endif
 #include <linux/task_io_accounting_ops.h>
 #include <linux/rcupdate.h>
 #include <linux/ptrace.h>
--- a/kernel/exit.c	2012-02-23 08:52:01.000000000 -0700
+++ b/kernel/exit.c	2012-02-23 09:48:34.000000000 -0700
@@ -38,6 +38,9 @@
 #include <linux/posix-timers.h>
 #include <linux/cn_proc.h>
 #include <linux/mutex.h>
+#ifdef __GENKSYMS__
+#include <linux/compat.h>
+#endif
 #include <linux/futex.h>
 #include <linux/pipe_fs_i.h>
 #include <linux/audit.h> /* for audit_free() */
