From: Ian Campbell <ijc@hellion.org.uk>
Date: Thu, 17 Jan 2013 08:55:21 +0000
Subject: Tweak gitignore for Debian pkg-kernel using git svn.

[bwh: Tweak further for pure git]

--- a/.gitignore
+++ b/.gitignore
@@ -39,8 +39,6 @@
 System.map
 Module.markers
 Module.symvers
-!.gitignore
-!.mailmap
 
 #
 # Generated include files
@@ -75,3 +73,10 @@ all.config
 *.orig
 *~
 \#*#
+
+#
+# Debian packaging: ignore everything at the top level, since it isn't
+# included in our repository
+#
+/*
+!/debian/
