From: Ben Hutchings <ben@decadent.org.uk>
Date: Thu, 9 Sep 2010 02:12:42 +0100
Subject: [PATCH] input: Avoid ABI change from compat support

---
--- a/drivers/input/input.c
+++ b/drivers/input/input.c
@@ -24,7 +24,9 @@
 #include <linux/mutex.h>
 #include <linux/rcupdate.h>
 #include <linux/smp_lock.h>
+#ifndef __GENKSYMS__
 #include "input-compat.h"
+#endif
 
 MODULE_AUTHOR("Vojtech Pavlik <vojtech@suse.cz>");
 MODULE_DESCRIPTION("Input core");
