From 7ba3254f3d5e474c6f7f775bb2440674e55a3e62 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Sat, 25 Sep 2010 01:17:59 +0100
Subject: [PATCH] mm: Avoid ABI change in 2.6.32.23

---
 include/linux/mmzone.h |   16 +++++++++-------
 1 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/include/linux/mmzone.h b/include/linux/mmzone.h
index 6c31a2a..9ae0d61 100644
--- a/include/linux/mmzone.h
+++ b/include/linux/mmzone.h
@@ -290,13 +290,6 @@ struct zone {
 	unsigned long watermark[NR_WMARK];
 
 	/*
-	 * When free pages are below this point, additional steps are taken
-	 * when reading the number of free pages to avoid per-cpu counter
-	 * drift allowing watermarks to be breached
-	 */
-	unsigned long percpu_drift_mark;
-
-	/*
 	 * We don't know if the memory that we're going to allocate will be freeable
 	 * or/and it will be released eventually, so to avoid totally wasting several
 	 * GB of ram we must reserve some of the lower zone memory (otherwise we risk
@@ -429,6 +422,15 @@ struct zone {
 	 * rarely used fields:
 	 */
 	const char		*name;
+
+#ifndef __GENKSYMS__
+	/*
+	 * When free pages are below this point, additional steps are taken
+	 * when reading the number of free pages to avoid per-cpu counter
+	 * drift allowing watermarks to be breached
+	 */
+	unsigned long percpu_drift_mark;
+#endif
 } ____cacheline_internodealigned_in_smp;
 
 typedef enum {
-- 
1.7.1

