From: Ben Hutchings <ben@decadent.org.uk>
Date: Sat, 08 Mar 2014 20:11:21 +0000
Subject: mm: Avoid ABI change in 3.2.55
Forwarded: not-needed

Extra #include, shut up genksyms.

--- a/mm/swap.c
+++ b/mm/swap.c
@@ -31,7 +31,9 @@
 #include <linux/backing-dev.h>
 #include <linux/memcontrol.h>
 #include <linux/gfp.h>
+#ifndef __GENKSYMS__
 #include <linux/hugetlb.h>
+#endif
 
 #include "internal.h"
 
