commit 613006cb46d1931bf7331fd3951efd7a0a0bb118
Author: dann frazier <dannf@debian.org>
Date:   Mon Aug 8 21:43:14 2011 -0600

    Avoid ABI change in fix for CVE-2011-2491

diff --git a/include/linux/sunrpc/sched.h b/include/linux/sunrpc/sched.h
index 67f63dd..b5e54f3 100644
--- a/include/linux/sunrpc/sched.h
+++ b/include/linux/sunrpc/sched.h
@@ -84,8 +84,10 @@ struct rpc_task {
 	long			tk_rtt;		/* round-trip time (jiffies) */
 
 	pid_t			tk_owner;	/* Process id for batching tasks */
-	unsigned char		tk_priority : 2,/* Task priority */
-				tk_rebind_retry : 2;
+	unsigned char		tk_priority : 2;/* Task priority */
+#ifndef __GENKSYMS__
+	unsigned char		tk_rebind_retry : 2;
+#endif
 #ifdef RPC_DEBUG
 	unsigned short		tk_pid;		/* debugging aid */
 #endif
