diff -urpN a/security/commoncap.c b/security/commoncap.c
--- a/security/commoncap.c	2012-05-05 11:06:19.843502544 -0600
+++ b/security/commoncap.c	2012-05-05 11:08:10.491502598 -0600
@@ -27,7 +27,9 @@
 #include <linux/sched.h>
 #include <linux/prctl.h>
 #include <linux/securebits.h>
+#ifndef __GENKSYMS__
 #include <linux/personality.h>
+#endif
 
 /*
  * If a non-root user executes a setuid-root binary in
