diff -urpN oom/include/linux/binfmts.h oom-noabi/include/linux/binfmts.h
--- oom/include/linux/binfmts.h	2010-12-28 19:07:39.746308959 -0700
+++ oom-noabi/include/linux/binfmts.h	2010-12-28 19:09:49.374136795 -0700
@@ -29,7 +29,6 @@ struct linux_binprm{
 	char buf[BINPRM_BUF_SIZE];
 #ifdef CONFIG_MMU
 	struct vm_area_struct *vma;
-	unsigned long vma_pages;
 #else
 # define MAX_ARG_PAGES	32
 	struct page *page[MAX_ARG_PAGES];
@@ -58,6 +57,11 @@ struct linux_binprm{
 	unsigned interp_flags;
 	unsigned interp_data;
 	unsigned long loader, exec;
+#ifndef __GENKSYMS__
+#ifdef CONFIG_MMU
+	unsigned long vma_pages;
+#endif
+#endif
 };
 
 #define BINPRM_FLAGS_ENFORCE_NONDUMP_BIT 0
