diff -urpN linux-source-2.6.26.orig/include/asm-x86/kvm_x86_emulate.h linux-source-2.6.26/include/asm-x86/kvm_x86_emulate.h
--- linux-source-2.6.26.orig/include/asm-x86/kvm_x86_emulate.h	2009-12-25 10:51:51.000000000 -0700
+++ linux-source-2.6.26/include/asm-x86/kvm_x86_emulate.h	2009-12-26 00:31:33.000000000 -0700
@@ -127,7 +127,7 @@ struct decode_cache {
 	unsigned long *override_base;
 	unsigned int d;
 	unsigned long regs[NR_VCPU_REGS];
-	unsigned long eip, eip_orig;
+	unsigned long eip;
 	/* modrm */
 	u8 modrm;
 	u8 modrm_mod;
@@ -138,6 +138,9 @@ struct decode_cache {
 	void *modrm_ptr;
 	unsigned long modrm_val;
 	struct fetch_cache fetch;
+#ifndef __GENKSYMS__
+	unsigned long eip_orig;
+#endif
 };
 
 struct x86_emulate_ctxt {
