diff -urpN a/net/socket.c b/net/socket.c
--- a/net/socket.c	2009-08-15 10:42:05.000000000 -0600
+++ b/net/socket.c	2009-08-15 11:09:49.000000000 -0600
@@ -698,6 +698,9 @@ static ssize_t sock_sendpage(struct file
 	if (more)
 		flags |= MSG_MORE;
 
+	if (!sock->ops->sendpage)
+		return sock_no_sendpage(sock, page, offset, size, flags);
+
 	return sock->ops->sendpage(sock, page, offset, size, flags);
 }
 
