--- a/include/linux/parport.h	2010-08-29 18:54:12.000000000 -0600
+++ b/include/linux/parport.h	2010-08-29 19:03:59.000000000 -0600
@@ -326,12 +326,14 @@ struct parport {
 	int spintime;
 	atomic_t ref_count;
 
+	struct list_head full_list;
+	struct parport *slaves[3];
+
+#ifndef __GENKSYMS__
 	unsigned long devflags;
 #define PARPORT_DEVPROC_REGISTERED	0
 	struct pardevice *proc_device;	/* Currently register proc device */
-
-	struct list_head full_list;
-	struct parport *slaves[3];
+#endif
 };
 
 #define DEFAULT_SPIN_TIME 500 /* us */
