diff -urpN a/include/linux/security.h b/include/linux/security.h
--- a/include/linux/security.h	2009-10-16 17:19:44.000000000 -0600
+++ b/include/linux/security.h	2009-10-16 17:23:16.000000000 -0600
@@ -28,7 +28,9 @@
 #include <linux/resource.h>
 #include <linux/sem.h>
 #include <linux/shm.h>
+#ifndef __GENKSYMS__
 #include <linux/mm.h> /* PAGE_ALIGN */
+#endif
 #include <linux/msg.h>
 #include <linux/sched.h>
 #include <linux/key.h>
