diff -urpN a/include/linux/security.h b/include/linux/security.h
--- a/include/linux/security.h	2010-01-31 22:07:58.000000000 -0700
+++ b/include/linux/security.h	2010-02-19 18:11:19.000000000 -0700
@@ -1994,7 +1994,7 @@ static inline int security_file_mmap (st
 				      unsigned long addr,
 				      unsigned long addr_only)
 {
-	return 0;
+	return cap_file_mmap(file, reqprot, prot, flags, addr, addr_only);
 }
 
 static inline int security_file_mprotect (struct vm_area_struct *vma,
@@ -2160,7 +2160,7 @@ static inline void security_task_to_inod
 static inline int security_ipc_permission (struct kern_ipc_perm *ipcp,
 					   short flag)
 {
-	return cap_file_mmap(file, reqprot, prot, flags, addr, addr_only);
+	return 0;
 }
 
 static inline int security_msg_msg_alloc (struct msg_msg * msg)
