From: Stefan <kdevel@vogtner.de>
Subject: tcp: Fix tcp_send_fin() regression in 2.6.32.66
Bug: https://bugzilla.kernel.org/show_bug.cgi?id=99161
Bug-Debian: https://bugs.debian.org/789037
Origin: https://bugzilla.kernel.org/attachment.cgi?id=178341&action=diff

--- a/net/ipv4/tcp_output.c	
+++ a/net/ipv4/tcp_output.c	
@@ -2136,7 +2136,7 @@ 
 	 */
 	if (tskb && (tcp_send_head(sk) || tcp_memory_pressure)) {
 coalesce:
-		TCP_SKB_CB(skb)->flags |= TCPCB_FLAG_FIN;
+		TCP_SKB_CB(tskb)->flags |= TCPCB_FLAG_FIN;
 		TCP_SKB_CB(tskb)->end_seq++;
 		tp->write_seq++;
 		if (!tcp_send_head(sk)) {
