diff -urpN linux-source-2.6.32.orig/net/ipv6/ip6_tunnel.c linux-source-2.6.32/net/ipv6/ip6_tunnel.c
--- linux-source-2.6.32.orig/net/ipv6/ip6_tunnel.c	2011-06-11 13:10:52.000000000 -0600
+++ linux-source-2.6.32/net/ipv6/ip6_tunnel.c	2011-07-21 00:23:41.002857909 -0600
@@ -1465,7 +1465,7 @@ static int __init ip6_tunnel_init(void)
 {
 	int  err;
 
-	err = register_pernet_device(&ip6_tnl_net_ops);
+	err = register_pernet_gen_device(&ip6_tnl_net_id, &ip6_tnl_net_ops);
 	if (err < 0)
 		goto out_pernet;
 
