diff -urpN a/arch/x86/ia32/ia32entry-xen.S b/arch/x86/ia32/ia32entry-xen.S
--- a/arch/x86/ia32/ia32entry-xen.S	2010-01-31 22:52:27.000000000 -0700
+++ b/arch/x86/ia32/ia32entry-xen.S	2010-01-31 22:55:26.000000000 -0700
@@ -537,7 +537,7 @@ ia32_sys_call_table:
 	.quad quiet_ni_syscall		/* streams2 */
 	.quad stub32_vfork            /* 190 */
 	.quad compat_sys_getrlimit
-	.quad sys32_mmap2
+	.quad sys_mmap_pgoff
 	.quad sys32_truncate64
 	.quad sys32_ftruncate64
 	.quad sys32_stat64		/* 195 */
