From 4f232537fcb3301b102f022b3dee8e294b462773 Mon Sep 17 00:00:00 2001
From: dann frazier <dannf@hp.com>
Date: Sun, 7 Mar 2010 19:47:12 -0700
Subject: [PATCH 4/4] 2.6.31.5/uvesafb-connector-disallow-unpliviged-users-to-send-netlink-packets.patch

---
 drivers/video/uvesafb.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/drivers/video/uvesafb.c b/drivers/video/uvesafb.c
index e945b64..b595d48 100644
--- a/drivers/video/uvesafb.c
+++ b/drivers/video/uvesafb.c
@@ -71,6 +71,9 @@ static void uvesafb_cn_callback(struct cn_msg *msg, struct netlink_skb_parms *ns
 	struct uvesafb_task *utask;
 	struct uvesafb_ktask *task;
 
+	if (!cap_raised(nsp->eff_cap, CAP_SYS_ADMIN))
+		return;
+
 	if (msg->seq >= UVESAFB_TASKS_MAX)
 		return;
 
-- 
1.7.0

