diff -urpN a/drivers/media/video/v4l2-compat-ioctl32.c b/drivers/media/video/v4l2-compat-ioctl32.c
--- a/drivers/media/video/v4l2-compat-ioctl32.c	2009-12-02 20:51:21.000000000 -0700
+++ b/drivers/media/video/v4l2-compat-ioctl32.c	2010-10-18 20:50:09.824588592 -0600
@@ -198,6 +198,7 @@ struct video_code32 {
 
 static int get_microcode32(struct video_code *kp, struct video_code32 __user *up)
 {
+	return -EFAULT;
 	if (!access_ok(VERIFY_READ, up, sizeof(struct video_code32)) ||
 		copy_from_user(kp->loadwhat, up->loadwhat, sizeof(up->loadwhat)) ||
 		get_user(kp->datasize, &up->datasize) ||
