commit 66f6339311a6f6c314219f74867799320e83a4fd
Author: dann frazier <dannf@debian.org>
Date:   Mon Sep 19 21:26:32 2011 -0600

    Remove unintentional call to perf_event_mmap()

diff --git a/mm/mmap.c b/mm/mmap.c
index acb42df..0b18249 100644
--- a/mm/mmap.c
+++ b/mm/mmap.c
@@ -1628,10 +1628,8 @@ int expand_upwards(struct vm_area_struct *vma, unsigned long address)
 		error = -ENOMEM;
 		if (vma->vm_pgoff + (size >> PAGE_SHIFT) >= vma->vm_pgoff) {
 			error = acct_stack_growth(vma, size, grow);
-			if (!error) {
+			if (!error)
 				vma->vm_end = address;
-				perf_event_mmap(vma);
-			}
 		}
 	}
 	anon_vma_unlock(vma);
