From: Ben Hutchings <ben@decadent.org.uk>
Date: Mon, 27 Sep 2010 12:40:23 +0100
Subject: [PATCH] xen: Fix typo in xen_percpu_chip definition

---
 drivers/xen/events.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/xen/events.c b/drivers/xen/events.c
index 688b211..373905b 100644
--- a/drivers/xen/events.c
+++ b/drivers/xen/events.c
@@ -959,7 +959,7 @@ static struct irq_chip xen_dynamic_chip __read_mostly = {
 	.retrigger	= retrigger_dynirq,
 };
 
-static struct irq_chip en_percpu_chip __read_mostly = {
+static struct irq_chip xen_percpu_chip __read_mostly = {
 	.name		= "xen-percpu",
 
 	.disable	= disable_dynirq,
-- 
1.7.1

