From: Nicolas Kaiser <nikai@nikai.net>
Date: Thu, 7 Oct 2010 23:29:27 +0000
Subject: [PATCH 172/207] net/tg3: simplify conditional

commit 2259dca36a2f0226596ea37132a46338fcff6722 upstream.

Simplify: ((a && !b) || (!a && b)) => (a != b)

Signed-off-by: Nicolas Kaiser <nikai@nikai.net>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 drivers/net/tg3.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/drivers/net/tg3.c b/drivers/net/tg3.c
index 03dfc94..9419104 100644
--- a/drivers/net/tg3.c
+++ b/drivers/net/tg3.c
@@ -9973,8 +9973,7 @@ static int tg3_set_pauseparam(struct net_device *dev, struct ethtool_pauseparam
 
 		if (!(phydev->supported & SUPPORTED_Pause) ||
 		    (!(phydev->supported & SUPPORTED_Asym_Pause) &&
-		     ((epause->rx_pause && !epause->tx_pause) ||
-		      (!epause->rx_pause && epause->tx_pause))))
+		     (epause->rx_pause != epause->tx_pause)))
 			return -EINVAL;
 
 		tp->link_config.flowctrl = 0;
-- 
1.7.5.4

