From: Francois Romieu <romieu@fr.zoreil.com>
Date: Fri, 24 Jun 2011 19:52:13 +0200
Subject: [PATCH 70/71] r8169: fix wrong register use.

commit 1e4e82baee8c2a8d753cbf6aa1a77326b71e59f0 upstream.

Signed-off-by: Francois Romieu <romieu@fr.zoreil.com>
Cc: Realtek linux nic maintainers <nic_swsd@realtek.com>
---
 drivers/net/r8169.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/r8169.c b/drivers/net/r8169.c
index cf8c1e8..14d8ce9 100644
--- a/drivers/net/r8169.c
+++ b/drivers/net/r8169.c
@@ -745,7 +745,7 @@ static void rtl8168_oob_notify(struct rtl8169_private *tp, u8 cmd)
 	msleep(2);
 	for (i = 0; i < 5; i++) {
 		udelay(100);
-		if (!(RTL_R32(ERIDR) & ERIAR_FLAG))
+		if (!(RTL_R32(ERIAR) & ERIAR_FLAG))
 			break;
 	}
 
-- 
1.7.5.4

