From: "J. R. Okajima" <hooanon05@yahoo.co.jp>
Date: Wed, 17 Apr 2013 23:31:04 +0900
Subject: [14/83] aufs: possible bugfix, make sure the target branch is upper
Origin: http://sourceforge.net/p/aufs/aufs3-standalone/ci/2cfb0493a9697d9f540b7bf607cca91f989cd950

Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>
---
 fs/aufs/file.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/fs/aufs/file.c
+++ b/fs/aufs/file.c
@@ -397,7 +397,8 @@ static int au_file_refresh_by_inode(stru
 	if (!S_ISDIR(inode->i_mode)
 	    && au_opt_test(au_mntflags(sb), PLINK)
 	    && au_plink_test(inode)
-	    && !d_unhashed(dentry)) {
+	    && !d_unhashed(dentry)
+	    && bstart < au_dbstart(dentry)) {
 		err = au_test_and_cpup_dirs(dentry, bstart);
 		if (unlikely(err))
 			goto out_unlock;
