From: stephen hemminger <shemminger@vyatta.com>
Date: Fri, 12 Feb 2010 06:58:07 +0000
Subject: [PATCH 37/56] sky2: version 1.27

commit cfc0861ba32643c63901832a583054bea0445961 upstream.

Signed-off-by: Stephen Hemminger <shemminger@vyatta.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 drivers/net/sky2.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/sky2.c b/drivers/net/sky2.c
index 4147530..dfff8e5 100644
--- a/drivers/net/sky2.c
+++ b/drivers/net/sky2.c
@@ -50,7 +50,7 @@
 #include "sky2.h"
 
 #define DRV_NAME		"sky2"
-#define DRV_VERSION		"1.26"
+#define DRV_VERSION		"1.27"
 #define PFX			DRV_NAME " "
 
 /*
-- 
1.7.1

