From: stephen hemminger <shemminger@vyatta.com>
Date: Thu, 13 May 2010 06:12:53 +0000
Subject: [PATCH 55/56] sky2: version 1.28

commit e0a67e2dd9aa01a59cd91a7b39a6af1282bbc214 upstream.

Version 1.28

Signed-off-by: Stephen Hemminger <shemminger@vyatta.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
[bwh: Adjust context for 2.6.32]
---
--- a/drivers/net/sky2.c
+++ b/drivers/net/sky2.c
@@ -53,7 +53,7 @@
 #include "sky2.h"
 
 #define DRV_NAME		"sky2"
-#define DRV_VERSION		"1.27"
+#define DRV_VERSION		"1.28"
 #define PFX			DRV_NAME " "
 
 /*
