--- a/arch/alpha/kernel/sys_titan.c	2007-07-08 16:32:17.000000000 -0700
+++ b/arch/alpha/kernel/sys_titan.c	2007-07-15 01:33:47.000000000 -0700
@@ -273,21 +273,23 @@
 static void __init
 titan_late_init(void)
 {
+	int retval;
+
 	/*
 	 * Enable the system error interrupts. These interrupts are 
 	 * all reported to the kernel as machine checks, so the handler
 	 * is a nop so it can be called to count the individual events.
 	 */
-	request_irq(63+16, titan_intr_nop, IRQF_DISABLED,
-		    "CChip Error", NULL);
-	request_irq(62+16, titan_intr_nop, IRQF_DISABLED,
-		    "PChip 0 H_Error", NULL);
-	request_irq(61+16, titan_intr_nop, IRQF_DISABLED,
-		    "PChip 1 H_Error", NULL);
-	request_irq(60+16, titan_intr_nop, IRQF_DISABLED,
-		    "PChip 0 C_Error", NULL);
-	request_irq(59+16, titan_intr_nop, IRQF_DISABLED,
-		    "PChip 1 C_Error", NULL);
+	retval = request_irq(63+16, titan_intr_nop, IRQF_DISABLED,
+			     "CChip Error", NULL);
+	retval = request_irq(62+16, titan_intr_nop, IRQF_DISABLED,
+			     "PChip 0 H_Error", NULL);
+	retval = request_irq(61+16, titan_intr_nop, IRQF_DISABLED,
+			     "PChip 1 H_Error", NULL);
+	retval = request_irq(60+16, titan_intr_nop, IRQF_DISABLED,
+			     "PChip 0 C_Error", NULL);
+	retval = request_irq(59+16, titan_intr_nop, IRQF_DISABLED,
+			     "PChip 1 C_Error", NULL);
 
 	/* 
 	 * Register our error handlers.
@@ -341,14 +343,16 @@
 static void __init
 privateer_init_pci(void)
 {
+	int retval;
+
 	/*
 	 * Hook a couple of extra err interrupts that the
 	 * common titan code won't.
 	 */
-	request_irq(53+16, titan_intr_nop, IRQF_DISABLED,
-		    "NMI", NULL);
-	request_irq(50+16, titan_intr_nop, IRQF_DISABLED,
-		    "Temperature Warning", NULL);
+	retval = request_irq(53+16, titan_intr_nop, IRQF_DISABLED,
+			     "NMI", NULL);
+	retval = request_irq(50+16, titan_intr_nop, IRQF_DISABLED,
+			     "Temperature Warning", NULL);
 
 	/*
 	 * Finish with the common version.
