From: Julien Cristau <jcristau@debian.org>
Date: Sun, 02 Oct 2016 17:52:02 +0000
Subject: hwrng: chaoskey - don't set quality field
Forwarded: not-needed
Bug-Debian: https://bugs.debian.org/839616

struct hwrng doesn't have that field in 3.16.

--- a/drivers/usb/misc/chaoskey.c
+++ b/drivers/usb/misc/chaoskey.c
@@ -216,19 +216,6 @@ static int chaoskey_probe(struct usb_int
 	dev->hwrng.name = dev->name ? dev->name : chaoskey_driver.name;
 	dev->hwrng.read = chaoskey_rng_read;
 
-	/* Set the 'quality' metric.  Quality is measured in units of
-	 * 1/1024's of a bit ("mills"). This should be set to 1024,
-	 * but there is a bug in the hwrng core which masks it with
-	 * 1023.
-	 *
-	 * The patch that has been merged to the crypto development
-	 * tree for that bug limits the value to 1024 at most, so by
-	 * setting this to 1024 + 1023, we get 1023 before the fix is
-	 * merged and 1024 afterwards. We'll patch this driver once
-	 * both bits of code are in the same tree.
-	 */
-	dev->hwrng.quality = 1024 + 1023;
-
 	dev->hwrng_registered = (hwrng_register(&dev->hwrng) == 0);
 	if (!dev->hwrng_registered)
 		usb_err(interface, "Unable to register with hwrng");
