--- /dev/null
+++ b/include/uapi/linux/netfilter/xt_BYPASS.h
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2019, AT&T Intellectual Property.  All rights reserved.
+ * Copyright (c) 2015 by Brocade Communications Systems, Inc.
+ * All rights reserved.
+ *
+ * "BYPASS" target extension for Xtables
+ */
+#ifndef _XT_BYPASS_TARGET_H
+#define _XT_BYPASS_TARGET_H
+
+struct xt_bypass_tginfo {
+	char iif[IFNAMSIZ];
+	char oif[IFNAMSIZ];
+	char raddr[INET6_ADDRSTRLEN];
+
+	struct xt_bypass_priv *priv __attribute__((aligned(8)));
+};
+
+#endif /* _XT_BYPASS_TARGET_H */
+
