From: Thomas Gleixner <tglx@linutronix.de>
Date: Mon, 14 Nov 2011 10:52:34 +0100
Subject: [PATCH 060/325] sysctl-include-fix.patch
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=4ce2f348ec73c1bf9c3d8109dff77cd5ec924b52

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/linux/sysctl.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/linux/sysctl.h b/include/linux/sysctl.h
index 703cfa33..b954c41 100644
--- a/include/linux/sysctl.h
+++ b/include/linux/sysctl.h
@@ -932,6 +932,7 @@ enum
 #include <linux/list.h>
 #include <linux/rcupdate.h>
 #include <linux/wait.h>
+#include <linux/atomic.h>
 
 /* For the /proc/sys support */
 struct ctl_table;
