From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 15 Jul 2011 16:24:45 +0200
Subject: [PATCH 124/397] net-ipv4-route-use-locks-on-up-rt.patch
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=e0f1d266e702456de89b516d6a80ecfe12959be5

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 net/ipv4/route.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/ipv4/route.c b/net/ipv4/route.c
index d3e45fcb7c33..f583bade4b14 100644
--- a/net/ipv4/route.c
+++ b/net/ipv4/route.c
@@ -254,7 +254,7 @@ struct rt_hash_bucket {
 };
 
 #if defined(CONFIG_SMP) || defined(CONFIG_DEBUG_SPINLOCK) || \
-	defined(CONFIG_PROVE_LOCKING)
+	defined(CONFIG_PROVE_LOCKING) || defined(CONFIG_PREEMPT_RT_FULL)
 /*
  * Instead of using one spinlock for each rt_hash_bucket, we use a table of spinlocks
  * The size of this table is a power of two and depends on the number of CPUS.
