From: Thomas Gleixner <tglx@linutronix.de>
Date: Sun, 17 Jul 2011 19:48:20 +0200
Subject: [PATCH 166/325] migrate-disable-rt-variant.patch
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=95ea4eb5aa8407b4223f9fa7d45e77966459af6b

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/linux/preempt.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/include/linux/preempt.h b/include/linux/preempt.h
index 8547d22..5db13ef 100644
--- a/include/linux/preempt.h
+++ b/include/linux/preempt.h
@@ -127,11 +127,15 @@ extern void migrate_enable(void);
 # define preempt_enable_rt()		preempt_enable()
 # define preempt_disable_nort()		do { } while (0)
 # define preempt_enable_nort()		do { } while (0)
+# define migrate_disable_rt()		migrate_disable()
+# define migrate_enable_rt()		migrate_enable()
 #else
 # define preempt_disable_rt()		do { } while (0)
 # define preempt_enable_rt()		do { } while (0)
 # define preempt_disable_nort()		preempt_disable()
 # define preempt_enable_nort()		preempt_enable()
+# define migrate_disable_rt()		do { } while (0)
+# define migrate_enable_rt()		do { } while (0)
 #endif
 
 #ifdef CONFIG_PREEMPT_NOTIFIERS
