From: Thomas Gleixner <tglx@linutronix.de>
Date: Sun, 17 Jul 2011 16:29:27 +0200
Subject: [PATCH 177/348] net-netif_rx_ni-migrate-disable.patch
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=8567e46ac31786ecf1a43d27ec445bcbb1f1c33f

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 net/core/dev.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index 1d41fdf..36a7af9 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -3033,11 +3033,11 @@ int netif_rx_ni(struct sk_buff *skb)
 {
 	int err;
 
-	preempt_disable();
+	migrate_disable();
 	err = netif_rx(skb);
 	if (local_softirq_pending())
 		thread_do_softirq();
-	preempt_enable();
+	migrate_enable();
 
 	return err;
 }
