From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 9 Jan 2013 23:03:29 +0100
Subject: [PATCH 293/397] sched: Init idle->on_rq in init_idle()
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=70c6f3791f961351e66669712cf98307c22400e8

Cc: stable-rt@vger.kernel.org
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Steven Rostedt <rostedt@goodmis.org>
---
 kernel/sched.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched.c b/kernel/sched.c
index 047d2b918244..5d3a2977a37a 100644
--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -6555,6 +6555,7 @@ void __cpuinit init_idle(struct task_struct *idle, int cpu)
 	rcu_read_unlock();
 
 	rq->curr = rq->idle = idle;
+	idle->on_rq = 1;
 #if defined(CONFIG_SMP)
 	idle->on_cpu = 1;
 #endif
