From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Tue, 7 May 2013 09:41:43 -0400
Subject: [PATCH 306/306] Linux 3.2.50-rt70 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=74f2e247603e39db748afb85904c5d4d7bb199c1

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..f36b5d4 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt70
