From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Tue, 7 May 2013 09:41:43 -0400
Subject: [PATCH 307/307] Linux 3.2.48-rt69 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=ffada7fe0e17da3f0dc0005f24ce3770ac1f4399

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..6518981 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt69
