From 6ec01c5506636e13e658e3c716b4d2dc3d55d7af Mon Sep 17 00:00:00 2001
From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Tue, 7 May 2013 09:41:43 -0400
Subject: [PATCH 308/308] Linux 3.2.44-rt65 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..e2eb197 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt65
