From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Wed, 11 Sep 2013 20:34:28 -0400
Subject: [PATCH 321/321] Linux 3.2.51-rt72 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=aa3bdb9e0eef8a012582800d31071015a5ee79e6

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..2c95a3c 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt72
