From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Tue, 11 Mar 2014 13:38:31 -0400
Subject: [PATCH 338/338] Linux 3.2.55-rt80 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=eaf7e8cc3d546e0766f754914820b760b8bed69c

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..5ba2c20 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt80
