From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Thu, 13 Mar 2014 06:34:41 -0400
Subject: [PATCH 341/341] Linux 3.2.55-rt81 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=7d8ae5ce6631dce8f50bc893e2c3419426717085

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..8269ec1 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt81
