From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Wed, 30 Apr 2014 12:28:35 -0400
Subject: [PATCH 343/343] Linux 3.2.60-rt87 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=e133f2bdf9f4c87e4c924780331a50b7a7c23b66

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..bd5aee3 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt87
