From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Thu, 30 Oct 2014 20:55:59 -0400
Subject: [PATCH 348/348] Linux 3.2.64-rt94 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=2d2093592b39d1baf9437055b8b6ad70c70962c4

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..8d02a9b 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt94
