From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Mon, 10 Aug 2015 10:14:45 -0400
Subject: [PATCH 374/374] Linux 3.2.71-rt104 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=877417cc2752999f1f75cdb0e0beb16fbd7b980e

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..d068aed 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt104
