From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Mon, 29 Feb 2016 12:01:30 -0500
Subject: [PATCH 380/380] Linux 3.2.77-rt111 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=3e2fab6ff342fc0b94d30cc90789df71f06a46c9

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a212663..9969a4b69fad 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt111
