From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Mon, 7 Mar 2016 15:31:17 -0500
Subject: [PATCH 386/386] Linux 3.2.78-rt113 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=878e0b19b385593d191f7459d16f131d7b9c56f1

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a212663..fdcd9167ca0b 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt113
