From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Thu, 14 Jul 2016 10:34:31 -0400
Subject: [PATCH 389/389] Linux 3.2.81-rt117 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=dd0c6a894695f5e24085721604d5aa683a3cd1ac

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a212663..9788245dd428 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt117
