From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Mon, 3 Oct 2016 17:59:59 -0400
Subject: [PATCH 395/395] Linux 3.2.82-rt119 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=5c1f0249d9621571cffb8c4d4d4fef41978b02d2

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a212663..ba6e8821d159 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt119
