From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Sun, 30 Oct 2016 12:41:19 -0400
Subject: [PATCH 397/397] Linux 3.2.84-rt122 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=97aa3579f8e498c4eb1c676a2bea32095082ce8f

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a212663..4de2714cd138 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt122
