From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Sun, 30 Oct 2016 12:41:19 -0400
Subject: [PATCH 397/397] Linux 3.2.89-rt127 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=ae6556a53e1a7caabb92b8fcfa3baab0873e7b80

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a212663..c85940d40a67 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt127
