From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Sun, 30 Oct 2016 12:41:19 -0400
Subject: [PATCH 397/397] Linux 3.2.95-rt133 REBASE
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=323a7359e7313359f1080b78d0a2cfe1d14988f7

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a212663..c2c7e0fb6685 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt133
