--- linux-2.6.15/include/asm-alpha/system.h.orig	2006-01-26 02:57:09.000000000 -0800
+++ linux-2.6.15/include/asm-alpha/system.h	2006-01-26 02:58:13.000000000 -0800
@@ -562,7 +562,7 @@
    if something tries to do an invalid cmpxchg().  */
 extern void __cmpxchg_called_with_bad_pointer(void);
 
-static inline unsigned long
+static inline __attribute__((always_inline)) unsigned long
 __cmpxchg(volatile void *ptr, unsigned long old, unsigned long new, int size)
 {
 	switch (size) {
