--- linux-2.6-2.6.15/drivers/ide/legacy/buddha.c.orig	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6-2.6.15/drivers/ide/legacy/buddha.c	2006-02-03 23:25:26.000000000 +0100
@@ -197,6 +197,7 @@
 			z_writeb(0, buddha_board+BUDDHA_IRQ_MR);
 		
 		for(i=0;i<buddha_num_hwifs;i++) {
+			memset(&hw, 0, sizeof(hw));
 			if(type != BOARD_XSURF) {
 				ide_setup_ports(&hw, (buddha_board+buddha_bases[i]),
 						buddha_offsets, 0,
