--- linux-2.6-2.6.15/arch/m68k/bvme6000/bvmeints.c.orig	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6-2.6.15/arch/m68k/bvme6000/bvmeints.c	2006-01-21 20:05:38.000000000 +0100
@@ -13,6 +13,7 @@
 
 #include <linux/types.h>
 #include <linux/kernel.h>
+#include <linux/interrupt.h>
 #include <linux/errno.h>
 #include <linux/seq_file.h>
 
--- linux-2.6-2.6.15/arch/m68k/mvme147/147ints.c.orig	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6-2.6.15/arch/m68k/mvme147/147ints.c	2006-01-21 20:06:45.000000000 +0100
@@ -13,6 +13,7 @@
 
 #include <linux/types.h>
 #include <linux/kernel.h>
+#include <linux/interrupt.h>
 #include <linux/errno.h>
 #include <linux/seq_file.h>
 
--- linux-2.6-2.6.15/arch/m68k/mvme16x/16xints.c.orig	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6-2.6.15/arch/m68k/mvme16x/16xints.c	2006-01-21 20:06:53.000000000 +0100
@@ -13,6 +13,7 @@
 
 #include <linux/types.h>
 #include <linux/kernel.h>
+#include <linux/interrupt.h>
 #include <linux/errno.h>
 #include <linux/seq_file.h>
 
