--- linux-2.6-2.6.15/drivers/video/cirrusfb.c.orig	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6-2.6.15/drivers/video/cirrusfb.c	2006-02-10 12:29:30.000000000 +0100
@@ -202,7 +202,7 @@
 		.sr07			= 0x20,
 		.sr07_1bpp		= 0x20,
 		.sr07_8bpp		= 0x21,
-		.sr1f			= 0x22
+		.sr1f			= 0x1e
 	},
 	[BT_SPECTRUM] = {
 		.name			= "CL Spectrum",
@@ -1229,7 +1229,7 @@
 
 		case BT_PICASSO:
 			DPRINTK ("(for Picasso)\n");
-			vga_wseq (regbase, CL_SEQR1F, 0x22);		/* ##vorher 22 MCLK select */
+			vga_wseq (regbase, CL_SEQR1F, 0x1e);	/* Peter Krumrich */ /* ##vorher 22 MCLK select */
 			vga_wseq (regbase, CL_SEQRF, 0xd0);	/* ## vorher d0 avoid FIFO underruns..? */
 			break;
 
@@ -1307,7 +1307,7 @@
 			break;
 
 		case BT_PICASSO:
-			vga_wseq (regbase, CL_SEQR1F, 0x22);		/* ### vorher 1c MCLK select */
+			vga_wseq (regbase, CL_SEQR1F, 0x1e);	/* Peter Krumrich */	/* ### vorher 1c MCLK select */
 			vga_wseq (regbase, CL_SEQRF, 0xb0);	/* Fast Page-Mode writes */
 			break;
 
@@ -1374,7 +1374,7 @@
 		case BT_PICASSO:
 			vga_wseq (regbase, CL_SEQR7, 0x27);
 			vga_wseq (regbase, CL_SEQRF, 0xb0);	/* Fast Page-Mode writes */
-			vga_wseq (regbase, CL_SEQR1F, 0x22);		/* MCLK select */
+			vga_wseq (regbase, CL_SEQR1F, 0x1e);	/* Peter Krumrich */ /* MCLK select */
 			break;
 
 		case BT_SPECTRUM:
@@ -1450,7 +1450,7 @@
 		case BT_PICASSO:
 			vga_wseq (regbase, CL_SEQR7, 0x25);
 			vga_wseq (regbase, CL_SEQRF, 0xb0);	/* Fast Page-Mode writes */
-			vga_wseq (regbase, CL_SEQR1F, 0x22);		/* MCLK select */
+			vga_wseq (regbase, CL_SEQR1F, 0x1e);	/* Peter Krumrich */ /* MCLK select */
 			break;
 
 		case BT_SPECTRUM:
