diff -u linux-kernel-2.6.12-2.6.12/arch/ppc/boot/Makefile linux-kernel-2.6.12-2.6.12/arch/ppc/boot/Makefile
--- linux-kernel-2.6.12-2.6.12/arch/ppc/boot/Makefile	2005-07-15 12:55:56.000000000 +0000
+++ linux-kernel-2.6.12-2.6.12/arch/ppc/boot/Makefile	2005-07-15 12:55:56.000000000 +0000
@@ -39,30 +39,30 @@
-	install -m 644 -o root -g root ld.script $(INSTALL_MKVMLINUZ)/boot
+	install -m 644 ld.script $(INSTALL_MKVMLINUZ)/boot
 	if [ -e openfirmware/note ]; then \
-		install -m 644 -o root -g root openfirmware/note $(INSTALL_MKVMLINUZ)/boot; \
+		install -m 644 openfirmware/note $(INSTALL_MKVMLINUZ)/boot; \
 	fi
 
 	mkdir -p $(INSTALL_MKVMLINUZ)/lib
-	install -m 644 -o root -g root ../../../lib/lib.a $(INSTALL_MKVMLINUZ)/lib
-	install -m 644 -o root -g root lib/lib.a $(INSTALL_MKVMLINUZ)/lib/ppc.a
-	install -m 644 -o root -g root common/lib.a $(INSTALL_MKVMLINUZ)/lib/common.a
+	install -m 644 ../../../lib/lib.a $(INSTALL_MKVMLINUZ)/lib
+	install -m 644 lib/lib.a $(INSTALL_MKVMLINUZ)/lib/ppc.a
+	install -m 644 common/lib.a $(INSTALL_MKVMLINUZ)/lib/common.a
 	if [ -e of1275/lib.a ]; then \
-		install -m 644 -o root -g root of1275/lib.a $(INSTALL_MKVMLINUZ)/lib/of.a; \
+		install -m 644 of1275/lib.a $(INSTALL_MKVMLINUZ)/lib/of.a; \
 	fi
 
 	if [ -e openfirmware/common.o ]; then \
 		mkdir -p $(INSTALL_MKVMLINUZ)/obj/openfirmware; \
-		install -m 644 -o root -g root openfirmware/*.o $(INSTALL_MKVMLINUZ)/obj/openfirmware; \
+		install -m 644 openfirmware/*.o $(INSTALL_MKVMLINUZ)/obj/openfirmware; \
 		rm -f $(INSTALL_MKVMLINUZ)/obj/openfirmware/image*; \
 	fi
 
 	mkdir -p $(INSTALL_MKVMLINUZ)/obj/simple
-	install -m 644 -o root -g root simple/*.o $(INSTALL_MKVMLINUZ)/obj/simple
+	install -m 644 simple/*.o $(INSTALL_MKVMLINUZ)/obj/simple
 	rm -f $(INSTALL_MKVMLINUZ)/obj/simple/image*.o
 
 	mkdir -p $(INSTALL_MKVMLINUZ)/utils
-	install -m 755 -o root -g root utils/addnote $(INSTALL_MKVMLINUZ)/utils
-	install -m 755 -o root -g root utils/hack-coff $(INSTALL_MKVMLINUZ)/utils
-	install -m 755 -o root -g root utils/mkbugboot $(INSTALL_MKVMLINUZ)/utils
-	install -m 755 -o root -g root utils/mknote $(INSTALL_MKVMLINUZ)/utils
-	install -m 755 -o root -g root utils/mkprep $(INSTALL_MKVMLINUZ)/utils
-	install -m 755 -o root -g root utils/mktree $(INSTALL_MKVMLINUZ)/utils
+	install -m 755 utils/addnote $(INSTALL_MKVMLINUZ)/utils
+	install -m 755 utils/hack-coff $(INSTALL_MKVMLINUZ)/utils
+	install -m 755 utils/mkbugboot $(INSTALL_MKVMLINUZ)/utils
+	install -m 755 utils/mknote $(INSTALL_MKVMLINUZ)/utils
+	install -m 755 utils/mkprep $(INSTALL_MKVMLINUZ)/utils
+	install -m 755 utils/mktree $(INSTALL_MKVMLINUZ)/utils
