--- linux-2.6.15/include/asm-s390/setup.h.orig	2006-02-01 11:27:00.000000000 +0100
+++ linux-2.6.15/include/asm-s390/setup.h	2006-02-01 11:31:54.000000000 +0100
@@ -67,6 +67,8 @@
 #define SET_CONSOLE_3215	do { console_mode = 2; } while (0)
 #define SET_CONSOLE_3270	do { console_mode = 3; } while (0)
 
+#ifdef __KERNEL__
+
 struct ipl_list_header {
 	u32 length;
 	u8  reserved[3];
@@ -114,6 +116,8 @@
 				 IPL_PARMBLOCK_ORIGIN)
 #define IPL_PARMBLOCK_SIZE	(IPL_PARMBLOCK_START->hdr.length)
 
+#endif /* __KERNEL__ */
+
 #else 
 
 #ifndef __s390x__
