diff --git a/include/linux/sched.h b/include/linux/sched.h
index 7ed3ae7..9f428dc 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1288,8 +1288,6 @@ struct task_struct {
 	atomic_t fs_excl;	/* holding fs exclusive resources */
 	struct rcu_head rcu;
 
-	struct list_head	*scm_work_list;
-
 	/*
 	 * cache last used pipe for splice
 	 */
@@ -1305,6 +1303,10 @@ struct task_struct {
 	int latency_record_count;
 	struct latency_record latency_record[LT_SAVECOUNT];
 #endif
+
+#ifndef __GENKSYMS__
+	struct list_head	*scm_work_list;
+#endif
 };
 
 /*
