diff --git a/Makefile b/Makefile
index 6a29b82..83813cc 100644
diff --git a/drivers/xen/events.c b/drivers/xen/events.c
index a4dc7bf..ac91a4e 100644
--- a/drivers/xen/events.c
+++ b/drivers/xen/events.c
@@ -930,7 +930,7 @@ static struct irq_chip xen_dynamic_chip __read_mostly = {
 	.retrigger	= retrigger_dynirq,
 };
 
-static struct irq_chip en_percpu_chip __read_mostly = {
+static struct irq_chip xen_percpu_chip __read_mostly = {
 	.name		= "xen-percpu",
 
 	.disable	= disable_dynirq,
