From: Ben Hutchings <ben@decadent.org.uk>
Date: Thu, 19 May 2011 04:29:25 +0100
Subject: [PATCH 2/6] PCI: Add dummy implementation of pci_dev_run_wake()

---
 include/linux/pci.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/include/linux/pci.h b/include/linux/pci.h
index e07d194..d67702a 100644
--- a/include/linux/pci.h
+++ b/include/linux/pci.h
@@ -745,6 +745,8 @@ int pci_wake_from_d3(struct pci_dev *dev, bool enable);
 pci_power_t pci_target_state(struct pci_dev *dev);
 int pci_prepare_to_sleep(struct pci_dev *dev);
 int pci_back_from_sleep(struct pci_dev *dev);
+/* Dummy for 2.6.32 */
+static inline bool pci_dev_run_wake(struct pci_dev *dev) { return false; }
 
 /* Functions for PCI Hotplug drivers to use */
 int pci_bus_find_capability(struct pci_bus *bus, unsigned int devfn, int cap);
-- 
1.7.5.4

