From: Ben Hutchings <ben@decadent.org.uk>
Date: Thu, 19 May 2011 04:48:31 +0100
Subject: [PATCH 3/6] pm: Define SET_SYSTEM_SLEEP_PM_OPS()

From commit d690b2cd222afc75320b9b8e9da7df02e9e630ca upstream.
---
 include/linux/pm.h |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/include/linux/pm.h b/include/linux/pm.h
index 3b7e04b..a0cc041 100644
--- a/include/linux/pm.h
+++ b/include/linux/pm.h
@@ -213,6 +213,14 @@ struct dev_pm_ops {
 	int (*runtime_idle)(struct device *dev);
 };
 
+#define SET_SYSTEM_SLEEP_PM_OPS(suspend_fn, resume_fn) \
+	.suspend = suspend_fn, \
+	.resume = resume_fn, \
+	.freeze = suspend_fn, \
+	.thaw = resume_fn, \
+	.poweroff = suspend_fn, \
+	.restore = resume_fn,
+
 /*
  * Use this if you want to use the same suspend and resume callbacks for suspend
  * to RAM and hibernation.
-- 
1.7.5.4

