From a24fc73896e9ff7d22a8a3a9e463a28124c60045 Mon Sep 17 00:00:00 2001
From: Sylvain Munaut <tnt@246tNt.com>
Date: Sun, 13 May 2007 15:57:21 +0200
Subject: [PATCH 17/21] powerpc: Fix typo in BestComm ATA task support code

Signed-off-by: Sylvain Munaut <tnt@246tNt.com>
---
 arch/powerpc/sysdev/bestcomm/ata.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/powerpc/sysdev/bestcomm/ata.c b/arch/powerpc/sysdev/bestcomm/ata.c
index 39efb9b..4b213b1 100644
--- a/arch/powerpc/sysdev/bestcomm/ata.c
+++ b/arch/powerpc/sysdev/bestcomm/ata.c
@@ -67,7 +67,7 @@ bcom_ata_init(int queue_len, int maxbufsize)
 
 	tsk->flags = BCOM_FLAGS_NONE;
 
-	bcom_ata_reset(tsk);
+	bcom_ata_reset_bd(tsk);
 
 	var = (struct bcom_ata_var *) bcom_task_var(tsk->tasknum);
 	inc = (struct bcom_ata_inc *) bcom_task_inc(tsk->tasknum);
-- 
1.5.1.2

