diff -r 832aac894efd arch/i386/kernel/vsyscall-note-xen.S
--- a/arch/i386/kernel/vsyscall-note-xen.S	Wed Nov 19 13:15:46 2008 +0000
+++ b/arch/i386/kernel/vsyscall-note-xen.S	Sun Nov 23 11:30:37 2008 +0100
@@ -17,7 +17,7 @@
  * like /lib/i686/tls/nosegneg.  Note that an /etc/ld.so.conf.d/file
  * corresponding to the bits here is needed to make ldconfig work right.
  * It should contain:
- *	hwcap 0 nosegneg
+ *	hwcap 1 nosegneg
  * to match the mapping of bit to name that we give here.
  */
 #define NOTE_KERNELCAP_BEGIN(ncaps, mask) \
@@ -26,7 +26,8 @@
 #define NOTE_KERNELCAP(bit, name) \
 	.byte bit; .asciz name
 #define NOTE_KERNELCAP_END ASM_ELF_NOTE_END
+#define VDSO_NOTE_NONEGSEG_BIT 1
 
-NOTE_KERNELCAP_BEGIN(1, 1)
-NOTE_KERNELCAP(1, "nosegneg")  /* Change 1 back to 0 when glibc is fixed! */
+NOTE_KERNELCAP_BEGIN(1, 1 << VDSO_NOTE_NONEGSEG_BIT)
+NOTE_KERNELCAP(VDSO_NOTE_NONEGSEG_BIT, "nosegneg")
 NOTE_KERNELCAP_END
