--- a/arch/i386/kernel/Makefile
+++ b/arch/i386/kernel/Makefile
@@ -94,1 +94,0 @@
-obj-y += fixup.o
--- a/arch/i386/kernel/entry-xen.S
+++ b/arch/i386/kernel/entry-xen.S
@@ -1205,7 +1205,0 @@
-ENTRY(fixup_4gb_segment)
-	RING0_INT_FRAME
-	pushl $do_fixup_4gb_segment
-	CFI_ADJUST_CFA_OFFSET 4
-	jmp error_code
-	CFI_ENDPROC
-
--- a/arch/i386/kernel/traps-xen.c
+++ b/arch/i386/kernel/traps-xen.c
@@ -101,8 +101,6 @@ asmlinkage void simd_coprocessor_error(void);
 asmlinkage void alignment_check(void);
 #ifndef CONFIG_XEN
 asmlinkage void spurious_interrupt_bug(void);
-#else
-asmlinkage void fixup_4gb_segment(void);
 #endif
 asmlinkage void machine_check(void);
 
@@ -1172,7 +1170,6 @@ static const trap_info_t __cpuinitconst trap_table[] = {
 	{ 12, 0, __KERNEL_CS, (unsigned long)stack_segment		},
 	{ 13, 0, __KERNEL_CS, (unsigned long)general_protection		},
 	{ 14, 0|4, __KERNEL_CS, (unsigned long)page_fault		},
-	{ 15, 0, __KERNEL_CS, (unsigned long)fixup_4gb_segment		},
 	{ 16, 0, __KERNEL_CS, (unsigned long)coprocessor_error		},
 	{ 17, 0, __KERNEL_CS, (unsigned long)alignment_check		},
 #ifdef CONFIG_X86_MCE
