diff -urpN a/arch/x86/kernel/process_64-xen.c b/arch/x86/kernel/process_64-xen.c
--- a/arch/x86/kernel/process_64-xen.c	2010-02-17 21:42:34.000000000 -0700
+++ b/arch/x86/kernel/process_64-xen.c	2010-02-17 21:44:00.000000000 -0700
@@ -779,6 +779,7 @@ void set_personality_ia32(void)
 
 	/* Make sure to be in 32bit mode */
 	set_thread_flag(TIF_IA32);
+	current->personality |= force_personality32;
 
 	/* Prepare the first "return" to user space */
 	current_thread_info()->status |= TS_COMPAT;
