diff -urpN a/arch/x86/kernel/process_32-xen.c b/arch/x86/kernel/process_32-xen.c
--- a/arch/x86/kernel/process_32-xen.c	2009-12-26 00:26:38.000000000 -0700
+++ b/arch/x86/kernel/process_32-xen.c	2009-12-26 00:46:54.000000000 -0700
@@ -153,7 +153,7 @@ void cpu_idle(void)
 
 	/* endless idle loop with no priority at all */
 	while (1) {
-		tick_nohz_stop_sched_tick();
+		tick_nohz_stop_sched_tick(1);
 		while (!need_resched()) {
 			void (*idle)(void);
 
diff -urpN a/arch/x86/kernel/process_64-xen.c b/arch/x86/kernel/process_64-xen.c
--- a/arch/x86/kernel/process_64-xen.c	2009-12-26 00:26:38.000000000 -0700
+++ b/arch/x86/kernel/process_64-xen.c	2009-12-26 00:46:47.000000000 -0700
@@ -146,7 +146,7 @@ void cpu_idle(void)
 	current_thread_info()->status |= TS_POLLING;
 	/* endless idle loop with no priority at all */
 	while (1) {
-		tick_nohz_stop_sched_tick();
+		tick_nohz_stop_sched_tick(1);
 		while (!need_resched()) {
 			void (*idle)(void);
 
