diff --git a/include/asm-x86/mach-xen/asm/pgtable.h b/include/asm-x86/mach-xen/asm/pgtable.h
index a9ff073..918c6c5 100644
--- a/include/asm-x86/mach-xen/asm/pgtable.h
+++ b/include/asm-x86/mach-xen/asm/pgtable.h
@@ -50,7 +50,7 @@
 /* If _PAGE_PRESENT is clear, we use these: */
 #define _PAGE_FILE	_PAGE_DIRTY	/* nonlinear file mapping,
 					 * saved PTE; unset:swap */
-#define _PAGE_PROTNONE	_PAGE_PSE	/* if the user mapped it with PROT_NONE;
+#define _PAGE_PROTNONE	_PAGE_UNUSED3	/* if the user mapped it with PROT_NONE;
 					   pte_present gives true */
 
 #ifndef __ASSEMBLY__
