From 7ed8586dd2ba2340a28a5d3d21f3f531a5bd30f8 Mon Sep 17 00:00:00 2001
From: Ian Campbell <ian.campbell@citrix.com>
Date: Tue, 5 Oct 2010 10:00:42 -0700
Subject: [PATCH] xen: grant table: do not truncate machine address on gnttab_copy_grant_page

Signed-off-by: Ian Campbell <ian.campbell@citrix.com>
Signed-off-by: Jeremy Fitzhardinge <jeremy.fitzhardinge@citrix.com>
---
 drivers/xen/grant-table.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/xen/grant-table.c b/drivers/xen/grant-table.c
index 236bf97..a23c569 100644
--- a/drivers/xen/grant-table.c
+++ b/drivers/xen/grant-table.c
@@ -548,7 +548,7 @@ int gnttab_copy_grant_page(grant_ref_t ref, struct page **pagep)
 	if (!xen_feature(XENFEAT_auto_translated_physmap)) {
 		set_phys_to_machine(page_to_pfn(new_page), INVALID_P2M_ENTRY);
 
-		mmu.ptr = (new_mfn << PAGE_SHIFT) | MMU_MACHPHYS_UPDATE;
+		mmu.ptr = PFN_PHYS(new_mfn) | MMU_MACHPHYS_UPDATE;
 		mmu.val = pfn;
 		err = HYPERVISOR_mmu_update(&mmu, 1, NULL, DOMID_SELF);
 		BUG_ON(err);
-- 
1.7.1

