netback: correct netbk_tx_err() to handle wrap around

Signed-off-by: Ian Campbell <ian.campbell@citrix.com>
Signed-off-by: Jan Beulich <JBeulich@suse.com>

--- a/drivers/xen/netback/netback.c
+++ b/drivers/xen/netback/netback.c
@@ -1011,7 +1011,7 @@ static void netbk_tx_err(netif_t *netif,
 
 	do {
 		make_tx_response(netif, txp, NETIF_RSP_ERROR);
-		if (cons >= end)
+		if (cons == end)
 			break;
 		txp = RING_GET_REQUEST(&netif->tx, cons++);
 	} while (1);
