From ff3483aef4dbbddf6ee5ca483555c0ef8f8a047f Mon Sep 17 00:00:00 2001
From: Vitaliy Gusev <vgusev@openvz.org>
Date: Thu, 25 Sep 2008 13:03:45 +0400
Subject: [PATCH] Fix erratum that causes memory corruption.

Signed-off-by: Vitaliy Gusev <vgusev@openvz.org>
Signed-off-by: Pavel Emelyanov <xemul@openvz.org>
---
 drivers/base/core.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/base/core.c b/drivers/base/core.c
index 660ecc0..47d5db2 100644
--- a/drivers/base/core.c
+++ b/drivers/base/core.c
@@ -1089,7 +1089,7 @@ EXPORT_SYMBOL_GPL(devices_init);
 
 void devices_fini(void)
 {
-	kset_unregister(devices_kset);
+	kset_unregister(ve_devices_kset);
 }
 EXPORT_SYMBOL_GPL(devices_fini);
 
-- 
1.6.0.6

