From 83ea78e92c2ef11d0b67cc5f2563f68679446886 Mon Sep 17 00:00:00 2001
From: Alexey Dobriyan <adobriyan@gmail.com>
Date: Wed, 15 Oct 2008 08:15:22 +0400
Subject: [PATCH] netns: fix net_generic array leak

Signed-off-by: Alexey Dobriyan <adobriyan@gmail.com>
Signed-off-by: Pavel Emelyanov <xemul@openvz.org>
---
 net/core/net_namespace.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/net/core/net_namespace.c b/net/core/net_namespace.c
index e1aeb0e..4ccdf17 100644
--- a/net/core/net_namespace.c
+++ b/net/core/net_namespace.c
@@ -102,7 +102,7 @@ static void net_free(struct net *net)
 		return;
 	}
 #endif
-
+	kfree(net->gen);
 	sysfs_completion = net->sysfs_completion;
 	kmem_cache_free(net_cachep, net);
 	if (sysfs_completion)
-- 
1.6.2

