diff -urpN a/kernel/cpt/cpt_process.c b/kernel/cpt/cpt_process.c
--- a/kernel/cpt/cpt_process.c	2010-02-09 12:02:40.000000000 -0700
+++ b/kernel/cpt/cpt_process.c	2010-02-09 12:13:10.000000000 -0700
@@ -941,7 +941,7 @@ static int dump_one_process(cpt_object_t
 	v->cpt_64bit = 0;
 #ifdef CONFIG_X86_64
 	/* Clear x86_64 specific flags */
-	v->cpt_thrflags &= ~(_TIF_FORK|_TIF_ABI_PENDING|_TIF_IA32);
+	v->cpt_thrflags &= ~(_TIF_FORK|_TIF_IA32);
 	if (!(task_thread_info(tsk)->flags & _TIF_IA32)) {
 		ctx->tasks64++;
 		v->cpt_64bit = 1;
