From: "David S. Miller" <davem@davemloft.net>
Date: Tue, 16 Mar 2010 23:36:24 -0700
Subject: [PATCH 064/166] e1000e: Fix build with CONFIG_PM disabled.

commit e50208a080a8abdfa6c85865362a1cf329e9fe31 upstream.

Signed-off-by: David S. Miller <davem@davemloft.net>
---
 drivers/net/e1000e/netdev.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/drivers/net/e1000e/netdev.c b/drivers/net/e1000e/netdev.c
index f57413b..2c337f7 100644
--- a/drivers/net/e1000e/netdev.c
+++ b/drivers/net/e1000e/netdev.c
@@ -5500,6 +5500,7 @@ static DEFINE_PCI_DEVICE_TABLE(e1000_pci_tbl) = {
 };
 MODULE_DEVICE_TABLE(pci, e1000_pci_tbl);
 
+#ifdef CONFIG_PM
 static const struct dev_pm_ops e1000_pm_ops = {
 	.suspend  = e1000_suspend,
 	.resume   = e1000_resume,
@@ -5511,6 +5512,7 @@ static const struct dev_pm_ops e1000_pm_ops = {
 	.runtime_resume = e1000_runtime_resume,
 	.runtime_idle = e1000_idle,
 };
+#endif
 
 /* PCI Device API Driver */
 static struct pci_driver e1000_driver = {
-- 
1.7.5.4

