From: Joe Perches <joe@perches.com>
Date: Sun, 14 Nov 2010 17:04:32 +0000
Subject: [PATCH 135/166] drivers/net/e1000e: Remove unnecessary semicolons

commit 1d51c4185bb369cb39ed8cc20b331508e47b35b2 upstream.

Signed-off-by: Joe Perches <joe@perches.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 drivers/net/e1000e/netdev.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/e1000e/netdev.c b/drivers/net/e1000e/netdev.c
index 694d096..94e22a1 100644
--- a/drivers/net/e1000e/netdev.c
+++ b/drivers/net/e1000e/netdev.c
@@ -4596,7 +4596,7 @@ dma_error:
 			i += tx_ring->count;
 		i--;
 		buffer_info = &tx_ring->buffer_info[i];
-		e1000_put_txbuf(adapter, buffer_info);;
+		e1000_put_txbuf(adapter, buffer_info);
 	}
 
 	return 0;
-- 
1.7.5.4

