diff -urpN linux-source-2.6.24.orig/arch/x86/kernel/entry_64.S linux-source-2.6.24/arch/x86/kernel/entry_64.S
--- linux-source-2.6.24.orig/arch/x86/kernel/entry_64.S	2008-01-24 15:58:37.000000000 -0700
+++ linux-source-2.6.24/arch/x86/kernel/entry_64.S	2008-06-03 01:26:47.000000000 -0600
@@ -779,7 +779,7 @@ paranoid_swapgs\trace:
 	swapgs
 paranoid_restore\trace:
 	RESTORE_ALL 8
-	iretq
+	jmp iret_label
 paranoid_userspace\trace:
 	GET_THREAD_INFO(%rcx)
 	movl threadinfo_flags(%rcx),%ebx
