diff -urpN linux-source-2.6.18.orig/fs/dnotify.c linux-source-2.6.18/fs/dnotify.c
--- linux-source-2.6.18.orig/fs/dnotify.c	2008-04-23 21:53:08.000000000 -0600
+++ linux-source-2.6.18/fs/dnotify.c	2008-05-02 18:33:30.000000000 -0600
@@ -69,6 +69,9 @@ int fcntl_dirnotify(int fd, struct file 
 	struct dnotify_struct **prev;
 	struct inode *inode;
 	fl_owner_t id = current->files;
+#ifndef __GENKSYMS__
+	struct file *f;
+#endif
 	int error = 0;
 
 	if ((arg & ~DN_MULTISHOT) == 0) {
@@ -96,8 +99,13 @@ int fcntl_dirnotify(int fd, struct file 
 	}
 
 #ifndef __GENKSYMS__
+	rcu_read_lock();
+	f = fcheck(fd);
+	rcu_read_unlock();
 	/* we'd lost the race with close(), sod off silently */
-	if (fcheck(fd) != filp)
+	/* note that inode->i_lock prevents reordering problems
+	 * between accesses to descriptor table and ->i_dnotify */
+	if (f != filp)
 		goto out_free;
 #endif
 
