diff -urpN linux-source-2.6.24.orig/fs/dnotify.c linux-source-2.6.24/fs/dnotify.c
--- linux-source-2.6.24.orig/fs/dnotify.c	2008-05-07 18:30:37.000000000 -0600
+++ linux-source-2.6.24/fs/dnotify.c	2008-05-07 18:32:12.000000000 -0600
@@ -20,7 +20,9 @@
 #include <linux/init.h>
 #include <linux/spinlock.h>
 #include <linux/slab.h>
+#ifndef __GENKSYMS__
 #include <linux/file.h>
+#endif
 
 int dir_notify_enable __read_mostly = 1;
 
@@ -97,11 +99,13 @@ int fcntl_dirnotify(int fd, struct file 
 	rcu_read_lock();
 	f = fcheck(fd);
 	rcu_read_unlock();
+#ifndef __GENKSYMS__
 	/* we'd lost the race with close(), sod off silently */
 	/* note that inode->i_lock prevents reordering problems
 	 * between accesses to descriptor table and ->i_dnotify */
 	if (f != filp)
 		goto out_free;
+#endif
 
 	error = __f_setown(filp, task_pid(current), PIDTYPE_PID, 0);
 	if (error)
