--- a/include/linux/sched.h	2008-11-18 22:50:17.000000000 -0700
+++ b/include/linux/sched.h	2008-11-18 23:08:15.000000000 -0700
@@ -997,7 +997,9 @@ struct task_struct {
 	struct task_delay_info *delays;
 #endif
 
+#ifndef __GENKSYMS__
 	struct list_head	*scm_work_list;
+#endif
 };
 
 static inline pid_t process_group(struct task_struct *tsk)
diff -urpN a/include/net/scm.h b/include/net/scm.h
--- a/include/net/scm.h	2008-12-10 00:18:25.000000000 -0700
+++ b/include/net/scm.h	2008-12-10 00:24:38.000000000 -0700
@@ -12,9 +12,11 @@
 
 struct scm_fp_list
 {
-	struct list_head	list;
 	int			count;
 	struct file		*fp[SCM_MAX_FD];
+#ifndef __GENKSYMS__
+	struct list_head	list;
+#endif
 };
 
 struct scm_cookie
