--- linux-source-2.6.18/fs/exec.c.orig	2006-09-19 21:42:06.000000000 -0600
+++ linux-source-2.6.18/fs/exec.c	2007-08-10 19:44:43.000000000 -0600
@@ -887,6 +887,7 @@
 	    file_permission(bprm->file, MAY_READ) ||
 	    (bprm->interp_flags & BINPRM_FLAGS_ENFORCE_NONDUMP)) {
 		suid_keys(current);
+		current->pdeath_signal = 0;
 		current->mm->dumpable = suid_dumpable;
 	}
 
@@ -977,8 +978,10 @@
 {
 	int unsafe;
 
-	if (bprm->e_uid != current->uid)
+	if (bprm->e_uid != current->uid) {
 		suid_keys(current);
+		current->pdeath_signal = 0;
+	}
 	exec_keys(current);
 
 	task_lock(current);
