From 9dfb7658ed3c31c986e5c724c1e4b9333ebf990c Mon Sep 17 00:00:00 2001
From: Geert Uytterhoeven <geert@linux-m68k.org>
Date: Mon, 17 Aug 2009 22:03:44 +0200
Subject: [PATCH] aranym: Update for v2.6.31 block layer changes

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>
---
 arch/m68k/emu/nfblock.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/m68k/emu/nfblock.c b/arch/m68k/emu/nfblock.c
index c131121..3d584f1 100644
--- a/arch/m68k/emu/nfblock.c
+++ b/arch/m68k/emu/nfblock.c
@@ -122,7 +122,7 @@ static int __init nfhd_init_one(int id, u32 blocks, u32 bsize)
 
 	dev->queue->queuedata = dev;
 	blk_queue_make_request(dev->queue, nfhd_make_request);
-	blk_queue_hardsect_size(dev->queue, bsize);
+	blk_queue_logical_block_size(dev->queue, bsize);
 
 	dev->disk = alloc_disk(16);
 	if (!dev->disk)
-- 
1.5.6.5

