From 507eeec8dadb53c77e52934c8c998541f7add9a9 Mon Sep 17 00:00:00 2001
From: Geert Uytterhoeven <geert@linux-m68k.org>
Date: Mon, 17 Aug 2009 22:04:29 +0200
Subject: [PATCH] aranym: Convert to net_device_ops

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>
Tested-by: Andreas Schwab <schwab@linux-m68k.org>
---
 arch/m68k/emu/nfeth.c |   18 +++++++++++++-----
 1 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/arch/m68k/emu/nfeth.c b/arch/m68k/emu/nfeth.c
index 5b118e9..f55dfd3 100644
--- a/arch/m68k/emu/nfeth.c
+++ b/arch/m68k/emu/nfeth.c
@@ -196,6 +196,17 @@ static struct net_device_stats *nfeth_get_stats(struct net_device *dev)
 	return &priv->stats;
 }
 
+static const struct net_device_ops nfeth_netdev_ops = {
+	.ndo_open		= nfeth_open,
+	.ndo_stop		= nfeth_stop,
+	.ndo_start_xmit		= nfeth_xmit,
+	.ndo_tx_timeout		= nfeth_tx_timeout,
+	.ndo_get_stats		= nfeth_get_stats,
+	.ndo_validate_addr	= eth_validate_addr,
+	.ndo_change_mtu		= eth_change_mtu,
+	.ndo_set_mac_address	= eth_mac_addr,
+};
+
 struct net_device * __init nfeth_probe(int unit)
 {
 	struct net_device *dev;
@@ -212,11 +223,8 @@ struct net_device * __init nfeth_probe(int unit)
 		return NULL;
 
 	dev->irq = nfEtherIRQ;
-	dev->open = nfeth_open;
-	dev->stop = nfeth_stop;
-	dev->hard_start_xmit = nfeth_xmit;
-	dev->tx_timeout = nfeth_tx_timeout;
-	dev->get_stats = nfeth_get_stats;
+	dev->netdev_ops = &nfeth_netdev_ops;
+
 	dev->flags |= NETIF_F_NO_CSUM;
 	memcpy(dev->dev_addr, mac, ETH_ALEN);
 
-- 
1.5.6.5

